org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java

```diff
@@ -0,0 +1,655 @@
+      if (!classFile.getElementName().contains("$")) {
+        filtered.add(classFile);
+      }
+    }
+    Object[] nonPackages = concatenate(filtered.toArray(), fragment.getNonJavaResources());
+    if (result.isEmpty()) {
+      return nonPackages;
+    }
+    Collections.addAll(result, nonPackages);
+    return result.toArray();
+  }
+
+  private static JarEntryFile findJarFile(JarEntryDirectory directory, String path) {
+    for (IJarEntryResource children : directory.getChildren()) {
+      if (children.isFile() && children.getFullPath().toOSString().equals(path)) {
+        return (JarEntryFile) children;
+      }
+      if (!children.isFile()) {
+        JarEntryFile file = findJarFile((JarEntryDirectory) children, path);
+        if (file != null) {
+          return file;
+        }
+      }
+    }
+    return null;
+  }
+
+  private static IPackageFragmentRoot getPackageFragmentRoot(
+      IJavaProject project, String id, IProgressMonitor pm) throws JavaModelException {
+    IPackageFragmentRoot[] roots = project.getAllPackageFragmentRoots();
+    IPackageFragmentRoot packageFragmentRoot = null;
+    for (IPackageFragmentRoot root : roots) {
+      if (pm.isCanceled()) {
+        throw new OperationCanceledException();
+      }
+      if (root.getHandleIdentifier().equals(id)) {
+        packageFragmentRoot = root;
+        break;
+      }
+    }
+    return packageFragmentRoot;
+  }
+
+  private static String readFileContent(JarEntryFile file) {
+    try (InputStream stream = (file.getContents())) {
+      return IOUtils.toString(stream, "UTF-8");
+    } catch (IOException | CoreException e) {
+      JavaLanguageServerPlugin.logException("Can't read file content: " + file.getFullPath(), e);
+    }
+    return null;
+  }
+
+  private static Object[] findJarDirectoryChildren(JarEntryDirectory directory, String path) {
+    String directoryPath = directory.getFullPath().toOSString();
+    if (directoryPath.equals(path)) {
+      return directory.getChildren();
+    }
+    if (path.startsWith(directoryPath)) {
+      for (IJarEntryResource resource : directory.getChildren()) {
+        String childrenPath = resource.getFullPath().toOSString();
+        if (childrenPath.equals(path)) {
+          return resource.getChildren();
+        }
+        if (path.startsWith(childrenPath) && resource instanceof JarEntryDirectory) {
+          findJarDirectoryChildren((JarEntryDirectory) resource, path);
+        }
+      }
+    }
+    return null;
+  }
+
+  private static Object[] getPackageFragmentRootContent(
+      IPackageFragmentRoot root, IProgressMonitor pm) throws JavaModelException {
+    ArrayList<Object> result = new ArrayList<>();
+    getHierarchicalPackageChildren(root, null, result, pm);
+    Object[] nonJavaResources = root.getNonJavaResources();
+    Collections.addAll(result, nonJavaResources);
+    return result.toArray();
+  }
+
+  /**
+   * Returns the hierarchical packages inside a given fragment or root.
```

Comment:
If I read this correctly, it this gives a tree of packages, with empty packages collapsed like so:

```
org.eclipse
  jdt.ls
  jdt.core
    IJavaElement.java
```
Since this is purely a display issue, I think in the long run we should move this concern to the client (browser) side.

Reply:
I think the description isn't so clear. Actually it is a tree where depth is 1 and collapsed default package (if it is a root of jar).In your example it will be:
``` 
org.eclipse
  jdt
```